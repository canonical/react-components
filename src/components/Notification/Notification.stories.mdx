import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";
import { useState } from "react";

import Notification, { NotificationSeverity } from "./Notification";

<Meta
  title="Notification"
  component={Notification}
  argTypes={{
    borderless: {
      control: {
        type: "boolean",
      },
    },
    children: {
      control: {
        type: "text",
      },
    },
    inline: {
      control: {
        type: "boolean",
      },
    },
    severity: {
      control: {
        type: "radio",
      },
      options: Object.values(NotificationSeverity),
    },
    timestamp: {
      control: {
        type: "text",
      },
    },
    title: {
      control: {
        type: "text",
      },
    },
  }}
  args={{
    borderless: false,
    inline: false,
    severity: NotificationSeverity.INFORMATION,
  }}
/>

export const Template = (args) => <Notification {...args} />;

### Notification

This is a [React](https://reactjs.org/) component for the Vanilla [Notification](https://docs.vanillaframework.io/patterns/notification/).

Notifications are used to display global information, our notification variants consist of default, caution, negative or positive.

The `NotificationProvider` component can be used to manage the notification state globally. Wrap your application components with it and then in any child component you can get the helper with `const notify = useNotify()` to easily trigger notifications:

- `notify.info("You should keep this in mind")`
- `notify.failure("Error when fooing a bar", errorObjectFromCatch)`
- `notify.success("The bar was foo'd", "Success")`

The `NotificationConsumer` component can be used to display notifications. It will automatically display any notifications that are added to the `NotificationProvider` state.

### Props

<ArgsTable of={Notification} />

#### Deprecated

The Notification API has changed since v0.18 in order to support new variants
and align with our component prop naming conventions.

| Deprecated prop names | Replaced by |
| --------------------- | ----------- |
| `type`                | `severity`  |
| `status`              | `title`     |
| `close`               | `onDismiss` |

The `notificationTypes` const has also been replaced with `NotificationSeverity`
to reflect the new prop name.

### Information

The information severity should be used to convey an information message.

<Canvas>
  <Story
    name="Information"
    args={{
      children: "Anyone with access can view your invited users.",
      severity: "information",
      title: "Permissions changed",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Caution

The caution severity should be used to convey information that is not critical but the user should be aware of.

<Canvas>
  <Story
    name="Caution"
    args={{
      children:
        "Custom storage configuration is only supported on Ubuntu, CentOS and RHEL.",
      severity: "caution",
      title: "Blocked",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Negative

The negative severity should be used to convey information that is critical and the user should take action.

<Canvas>
  <Story
    name="Negative"
    args={{
      children: "Node must be connected to a network.",
      severity: "negative",
      title: "Error",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Positive

The positive severity should be used to convey success or completion.

<Canvas>
  <Story
    name="Positive"
    args={{
      children: "Code successfully reformatted.",
      severity: "positive",
      title: "Success",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Inline

When vertical space is limited, you can use the inline variant.

<Canvas>
  <Story
    name="Inline"
    args={{
      children: "Not enough space.",
      inline: true,
      severity: "negative",
      title: "Error:",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Borderless

In cases where a notification sits inside another component, such as a table cell or a card, it may be useful to remove the outer border and highlight bar.

<Canvas>
  <Story
    name="Borderless"
    args={{
      borderless: true,
      children: "Only 8GB storage remaining.",
      severity: "caution",
      title: "Warning",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Actions

Actions will appear below the notification message.

<Canvas>
  <Story
    name="Actions"
    args={{
      actions: [
        { label: "Action 1", onClick: () => null },
        { label: "Action 2", onClick: () => null },
      ],
      children:
        "Body lorem ipsum dolor sit amet consequiteor. Lorem ipsum dolor sit amet consequiteor.",
      severity: "information",
      title: "Title",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Dismissible

Notifications that can be dismissed should be given an onDismiss function.

<Canvas>
  <Story name="Dismissible">
    {() => {
      const [show, setShow] = useState(true);
      if (!show) {
        return null;
      }
      return (
        <Notification onDismiss={() => setShow(false)} title="Dismissible">
          Click the dismiss button to get rid of me.
        </Notification>
      );
    }}
  </Story>
</Canvas>

### Timeout

Notifications can be automatically dismissed after a timeout interval.

<Canvas>
  <Story name="Timeout">
    {() => {
      const [show, setShow] = useState(true);
      const [timer, setTimer] = useState(0);
      if (!show) {
        return null;
      }
      return (
        <Notification
          actions={[{ label: "Start timer", onClick: () => setTimer(3000) }]}
          onDismiss={() => setShow(false)}
          timeout={timer}
          title="Timeout"
        >
          Click the action to start a 3 second timer.
        </Notification>
      );
    }}
  </Story>
</Canvas>

### Timestamp

For notifications in which recency is important, you can include a section for time.

<Canvas>
  <Story
    name="Timestamp"
    args={{
      children:
        "Body lorem ipsum dolor sit amet consequiteor. Lorem ipsum dolor sit amet consequiteor.",
      severity: "information",
      timestamp: "1h ago",
      title: "Title",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
