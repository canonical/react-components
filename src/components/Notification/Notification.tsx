import classNames from "classnames";
import React, { ElementType, useEffect, useRef } from "react";
import type { HTMLProps, ReactNode } from "react";

import Button, { ButtonAppearance } from "../Button";
import { IS_DEV } from "../../utils";

import type { ClassName, PropsWithSpread, ValueOf } from "types";

export enum Label {
  Close = "Close notification",
}

export const NotificationSeverity = {
  CAUTION: "caution",
  INFORMATION: "information",
  NEGATIVE: "negative",
  POSITIVE: "positive",
} as const;

export const DefaultTitles = {
  [NotificationSeverity.CAUTION]: "Warning",
  [NotificationSeverity.INFORMATION]: "Info",
  [NotificationSeverity.NEGATIVE]: "Error",
  [NotificationSeverity.POSITIVE]: "Success",
};

type NotificationAction = {
  label: string;
  onClick: () => void;
};

/**
 * The props for the Notification component.
 */
export type Props = PropsWithSpread<
  {
    /**
     * A list of up to two actions that the notification can perform.
     */
    actions?: NotificationAction[];
    /**
     * Whether the notification should not have a border.
     */
    borderless?: boolean;
    /**
     * The notification message content.
     */
    children?: ReactNode;
    /**
     * Optional class(es) to apply to the parent notification element.
     */
    className?: ClassName;
    /**
     * **Deprecated**. Use `onDismiss` instead.
     */
    close?: never;
    /**
     * Whether the title should display inline with the message.
     */
    inline?: boolean;
    /**
     * The function to run when dismissing/closing the notification.
     */
    onDismiss?: () => void;
    /**
     * The severity of the notification.
     */
    severity?: ValueOf<typeof NotificationSeverity>;
    /**
     * **Deprecated**. Use `title` instead.
     */
    status?: never;
    /**
     * The amount of time (in ms) until the notification is automatically dismissed.
     */
    timeout?: number;
    /**
     * A relevant timestamp for the notification, e.g. when it was created.
     */
    timestamp?: ReactNode;
    /**
     * The title of the notification.
     */
    title?: ReactNode;
    /**
     * Optional element or component to use for the title.
     */
    titleElement?: ElementType;
    /**
     * **Deprecated**. Use `severity` instead.
     */
    type?: never;
  },
  HTMLProps<HTMLDivElement>
>;

/**
This is a [React](https://reactjs.org/) component for the Vanilla [Notification](https://docs.vanillaframework.io/patterns/notification/).

Notifications are used to display global information, our notification variants consist of default, caution, negative or positive.

The `NotificationProvider` component can be used to manage the notification state globally. Wrap your application components with it and then in any child component you can get the helper with `const notify = useNotify()` to easily trigger notifications:

* `notify.info("You should keep this in mind")`
* `notify.failure("Error when fooing a bar", errorObjectFromCatch)`
* `notify.success("The bar was foo'd", "Success")`

The `NotificationConsumer` component can be used to display notifications. It will automatically display any notifications that are added to the `NotificationProvider` state.

The Notification API has changed since v0.18 in order to support new variants
and align with our component prop naming conventions.

| Deprecated prop names | Replaced by |
| --------------------- | ----------- |
| `type`                | `severity`  |
| `status`              | `title`     |
| `close`               | `onDismiss` |

The `notificationTypes` const has also been replaced with `NotificationSeverity`
to reflect the new prop name.
 */
const Notification = ({
  actions,
  borderless = false,
  children,
  className,
  close,
  inline = false,
  onDismiss,
  severity = NotificationSeverity.INFORMATION,
  status,
  timeout,
  timestamp,
  title,
  titleElement: TitleComponent = "h5",
  type,
  ...props
}: Props): JSX.Element => {
  const timeoutId = useRef(null);
  const hasActions = actions?.length > 0;
  const showMeta = !!timestamp || hasActions;

  useEffect(() => {
    if (timeout && onDismiss) {
      timeoutId.current = setTimeout(() => onDismiss(), timeout);
    }
    return () => clearTimeout(timeoutId.current);
  }, [onDismiss, timeout]);

  if (IS_DEV && (close || status || type)) {
    console.warn(
      "The Notification component is using deprecated props. Refer to the deprecated list for details: https://canonical.github.io/react-components/?path=/docs/notification--information#deprecated"
    );
  }

  return (
    <div
      className={classNames(className, {
        [`p-notification--${severity}`]: !!severity,
        "p-notification": !severity,
        "is-borderless": borderless,
        "is-inline": inline,
      })}
      {...props}
    >
      <div className="p-notification__content">
        {title && (
          <TitleComponent
            className="p-notification__title"
            data-testid="notification-title"
          >
            {title}
          </TitleComponent>
        )}
        {inline && <>&ensp;</>}
        <p className="p-notification__message">{children}</p>
        {onDismiss && (
          <button
            className="p-notification__close"
            data-testid="notification-close-button"
            onClick={onDismiss}
          >
            {Label.Close}
          </button>
        )}
      </div>
      {showMeta && (
        <div className="p-notification__meta" data-testid="notification-meta">
          {timestamp && (
            <span
              className="p-notification__timestamp"
              data-testid="notification-timestamp"
            >
              {timestamp}
            </span>
          )}
          {hasActions ? (
            <div className="p-notification__actions">
              {actions.map((action, i) => (
                <Button
                  appearance={ButtonAppearance.LINK}
                  className="p-notification__action"
                  data-testid="notification-action"
                  key={`${action.label}-${i}`}
                  onClick={action.onClick}
                >
                  {action.label}
                </Button>
              ))}
            </div>
          ) : null}
        </div>
      )}
    </div>
  );
};

export default Notification;
