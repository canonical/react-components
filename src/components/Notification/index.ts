export { default, NotificationSeverity } from "./Notification";
export type { Props as NotificationProps } from "./Notification";
export type { ToastNotificationType } from "./ToastNotification";
export {
  ToastNotification,
  ToastNotificationList,
  useToastNotification,
  ToastNotificationProvider,
} from "./ToastNotification";
