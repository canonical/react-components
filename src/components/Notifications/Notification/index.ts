export { default, NotificationSeverity } from "./Notification";
export type { Props as NotificationProps } from "./Notification";
