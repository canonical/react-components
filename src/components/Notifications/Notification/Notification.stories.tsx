import React from "react";
import { useState } from "react";
import { Meta, StoryObj } from "@storybook/react";

import Notification, { NotificationSeverity } from "./Notification";

const meta: Meta<typeof Notification> = {
  component: Notification,
  tags: ["autodocs"],

  argTypes: {
    borderless: {
      control: {
        type: "boolean",
      },
    },

    children: {
      control: {
        type: "text",
      },
    },

    inline: {
      control: {
        type: "boolean",
      },
    },

    severity: {
      control: {
        type: "radio",
      },

      options: Object.values(NotificationSeverity),
    },

    timestamp: {
      control: {
        type: "text",
      },
    },

    title: {
      control: {
        type: "text",
      },
    },
  },

  args: {
    borderless: false,
    inline: false,
    severity: NotificationSeverity.INFORMATION,
  },
};

export default meta;

type Story = StoryObj<typeof Notification>;

/**
 * The information severity should be used to convey an information message.
 */
export const Information: Story = {
  name: "Information",

  args: {
    children: "Anyone with access can view your invited users.",
    severity: "information",
    title: "Permissions changed",
  },
};

/**
 * The caution severity should be used to convey information that is not critical but the user should be aware of.
 */
export const Caution: Story = {
  name: "Caution",

  args: {
    children:
      "Custom storage configuration is only supported on Ubuntu, CentOS and RHEL.",
    severity: "caution",
    title: "Blocked",
  },
};

/**
 * The negative severity should be used to convey information that is critical and the user should take action.
 */
export const Negative: Story = {
  name: "Negative",

  args: {
    children: "Node must be connected to a network.",
    severity: "negative",
    title: "Error",
  },
};

/**
 * The positive severity should be used to convey success or completion.
 */
export const Positive: Story = {
  name: "Positive",

  args: {
    children: "Code successfully reformatted.",
    severity: "positive",
    title: "Success",
  },
};

/**
 * When vertical space is limited, you can use the inline variant.
 */
export const Inline: Story = {
  name: "Inline",

  args: {
    children: "Not enough space.",
    inline: true,
    severity: "negative",
    title: "Error:",
  },
};

/**
 * In cases where a notification sits inside another component, such as a table cell or a card, it may be useful to remove the outer border and highlight bar.
 */
export const Borderless: Story = {
  name: "Borderless",

  args: {
    borderless: true,
    children: "Only 8GB storage remaining.",
    severity: "caution",
    title: "Warning",
  },
};

/**
 * Actions will appear below the notification message.
 */
export const Actions: Story = {
  name: "Actions",

  args: {
    actions: [
      {
        label: "Action 1",
        onClick: () => null,
      },
      {
        label: "Action 2",
        onClick: () => null,
      },
    ],

    children:
      "Body lorem ipsum dolor sit amet consequiteor. Lorem ipsum dolor sit amet consequiteor.",
    severity: "information",
    title: "Title",
  },
};

/**
 * Notifications that can be dismissed should be given an onDismiss function.
 */
export const Dismissible: Story = {
  render: () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [show, setShow] = useState(true);

    if (!show) {
      return null;
    }

    return (
      <Notification onDismiss={() => setShow(false)} title="Dismissible">
        Click the dismiss button to get rid of me.
      </Notification>
    );
  },

  name: "Dismissible",
};

/**
 * Notifications can be automatically dismissed after a timeout interval.
 */
export const Timeout: Story = {
  render: () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [show, setShow] = useState(true);
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [timer, setTimer] = useState(0);

    if (!show) {
      return null;
    }

    return (
      <Notification
        actions={[
          {
            label: "Start timer",
            onClick: () => setTimer(3000),
          },
        ]}
        onDismiss={() => setShow(false)}
        timeout={timer}
        title="Timeout"
      >
        Click the action to start a 3 second timer.
      </Notification>
    );
  },

  name: "Timeout",
};

/**
 * For notifications in which recency is important, you can include a section for time.
 */
export const Timestamp: Story = {
  name: "Timestamp",

  args: {
    children:
      "Body lorem ipsum dolor sit amet consequiteor. Lorem ipsum dolor sit amet consequiteor.",
    severity: "information",
    timestamp: "1h ago",
    title: "Title",
  },
};
