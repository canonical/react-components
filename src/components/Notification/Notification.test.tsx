import { mount, shallow } from "enzyme";
import React from "react";

import Notification, { NotificationSeverity } from "./Notification";

describe("Notification", () => {
  beforeEach(() => {
    jest.useFakeTimers("modern");
  });

  it("renders", () => {
    const wrapper = shallow(
      <Notification
        severity={NotificationSeverity.INFORMATION}
        title="Permissions changed"
      >
        Anyone with access can view your invited users.
      </Notification>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("can add additional classes", () => {
    const wrapper = shallow(<Notification className="extra-class" />);
    const className = wrapper.prop("className");
    expect(className.includes("extra-class")).toBe(true);
  });

  it("can be given a title", () => {
    const wrapper = shallow(<Notification />);
    expect(wrapper.find("[data-testid='notification-title']").exists()).toBe(
      false
    );

    wrapper.setProps({ title: "Title" });
    expect(wrapper.find("[data-testid='notification-title']").exists()).toBe(
      true
    );
  });

  it("can be made inline", () => {
    const wrapper = shallow(<Notification />);
    expect(wrapper.prop("className").includes("is-inline")).toBe(false);

    wrapper.setProps({ inline: true });
    expect(wrapper.prop("className").includes("is-inline")).toBe(true);
  });

  it("can have a borderless appearance", () => {
    const wrapper = shallow(<Notification />);
    expect(wrapper.prop("className").includes("is-borderless")).toBe(false);

    wrapper.setProps({ borderless: true });
    expect(wrapper.prop("className").includes("is-borderless")).toBe(true);
  });

  it("can be made dismissible", () => {
    const onDismiss = jest.fn();
    const wrapper = shallow(<Notification />);
    expect(
      wrapper.find("[data-testid='notification-close-button']").exists()
    ).toBe(false);

    wrapper.setProps({ onDismiss });
    expect(
      wrapper.find("[data-testid='notification-close-button']").exists()
    ).toBe(true);

    wrapper.find("[data-testid='notification-close-button']").simulate("click");
    expect(onDismiss).toHaveBeenCalled();
  });

  it("can be given a timestamp", () => {
    const wrapper = shallow(<Notification />);
    expect(wrapper.find("[data-testid='notification-meta']").exists()).toBe(
      false
    );
    expect(
      wrapper.find("[data-testid='notification-timestamp']").exists()
    ).toBe(false);

    wrapper.setProps({ timestamp: "1h ago" });
    expect(wrapper.find("[data-testid='notification-meta']").exists()).toBe(
      true
    );
    expect(
      wrapper.find("[data-testid='notification-timestamp']").exists()
    ).toBe(true);
  });

  it("can be given actions", () => {
    const onActionClick = jest.fn();
    const wrapper = shallow(<Notification />);
    expect(wrapper.find("[data-testid='notification-meta']").exists()).toBe(
      false
    );
    expect(wrapper.find("[data-testid='notification-action']").exists()).toBe(
      false
    );

    wrapper.setProps({
      actions: [{ label: "Action", onClick: onActionClick }],
    });
    expect(wrapper.find("[data-testid='notification-meta']").exists()).toBe(
      true
    );
    expect(wrapper.find("[data-testid='notification-action']").exists()).toBe(
      true
    );

    wrapper.find("[data-testid='notification-action']").simulate("click");
    expect(onActionClick).toHaveBeenCalled();
  });

  it("can automatically dismiss the notification after a given timeout period", () => {
    const onDismiss = jest.fn();
    const timeout = 1000;
    mount(<Notification onDismiss={onDismiss} timeout={timeout} />);
    expect(onDismiss).not.toHaveBeenCalled();

    jest.advanceTimersByTime(timeout + 1);
    expect(onDismiss).toHaveBeenCalled();
  });
});
