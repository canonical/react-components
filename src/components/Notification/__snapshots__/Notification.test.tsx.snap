// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders 1`] = `
<div
  className="p-notification--information"
>
  <div
    className="p-notification__content"
  >
    <h5
      className="p-notification__title"
      data-testid="notification-title"
    >
      Permissions changed
    </h5>
    <p
      className="p-notification__message"
    >
      Anyone with access can view your invited users.
    </p>
  </div>
</div>
`;
